nameko/testing/pytest.py

```diff
@@ -84,30 +85,51 @@ def rabbit_manager(request):
 
 @pytest.yield_fixture()
 def rabbit_config(request, rabbit_manager):
+    import random
+    import string
     from kombu import pools
-    from nameko.testing.utils import (
-        reset_rabbit_vhost, reset_rabbit_connections,
-        get_rabbit_connections, get_rabbit_config)
+    from six.moves.urllib.parse import urlparse  # pylint: disable=E0611
+    from nameko.testing.utils import get_rabbit_connections
 
     amqp_uri = request.config.getoption('AMQP_URI')
 
-    conf = get_rabbit_config(amqp_uri)
+    uri = urlparse(amqp_uri)
+    username = uri.username
+    vhost = uri.path[1:]
 
-    reset_rabbit_connections(conf['vhost'], rabbit_manager)
-    reset_rabbit_vhost(conf['vhost'], conf['username'], rabbit_manager)
+    use_random_vost = (vhost == ":random:")
+
+    if use_random_vost:
+        vhost = "test_{}".format(
+            "".join(random.choice(string.ascii_lowercase) for _ in range(10))
+        )
+        amqp_uri = "{}://{}/{}".format(uri.scheme, uri.netloc, vhost)
+        rabbit_manager.create_vhost(vhost)
+        rabbit_manager.set_vhost_permissions(vhost, username, '.*', '.*', '.*')
+
+    conf = {
+        'AMQP_URI': amqp_uri,
+        'username': username,
+        'vhost': vhost
+    }
 
     yield conf
 
     pools.reset()  # close connections in pools
 
     # raise a runtime error if the test leaves any connections lying around
-    connections = get_rabbit_connections(conf['vhost'], rabbit_manager)
-    open_connections = [
-        conn for conn in connections if conn['state'] != "closed"
-    ]
-    if open_connections:
-        count = len(open_connections)
-        raise RuntimeError("{} rabbit connection(s) left open.".format(count))
+    try:
```

Comment:
I think you've misunderstood how try/finally flow control works. I'm not handling the ``RuntimeError`` here.

I want to raise the exception if any connections are left over. This exception will bubble up and stop the execution of the test. The ``finally`` is there because, whether I throw the exception or not, I should always delete any vhost the test created. In the case where the user provided a vhost (i.e. it's not a random one) we assume they don't want it destroyed.

Regarding your other points:

* Yep, agree that the error should identify the connection. I will add.
* Checking ``conn['state'] == "open"`` isn't sufficient because there are other states that should also be considered "lying around"
* ``rabbit_manager.get_connections`` gets connections on all vhosts. ``get_rabbit_connections`` filters that response. 
* On ``RuntimeError`` vs something else... was here before this PR but I'm happy to raise something more specific if we can think of something. Maybe ``CleanupError``?